File: python/src/epostak/resources/white_label.py
"""White Label participant registration and migration lifecycle."""

from __future__ import annotations

from typing import Optional, TYPE_CHECKING
from urllib.parse import quote

from epostak.resources.documents import _BaseResource, _build_query, _idempotency_headers

if TYPE_CHECKING:
    from epostak.types import (
        WhiteLabelMigrationCodeResponse,
        WhiteLabelParticipant,
        WhiteLabelParticipantList,
        WhiteLabelParticipantMigrationRequest,
        WhiteLabelParticipantOperation,
        WhiteLabelParticipantRegistrationRequest,
    )


def _white_label_idempotency_key(value: str) -> str:
    if not isinstance(value, str) or not value.strip() or len(value.encode("utf-8")) > 255:
        raise ValueError("White Label idempotency key must be 1-255 UTF-8 bytes")
    return value


class WhiteLabelResource(_BaseResource):
    """Integrator-scoped White Label participant lifecycle.

    Calls never send ``X-Firm-Id``. The authenticated White Label integrator
    determines participant ownership on the server.
    """

    def list_participants(
        self,
        *,
        limit: Optional[int] = None,
        cursor: Optional[str] = None,
    ) -> WhiteLabelParticipantList:
        return self._request(
            "GET",
            "/white-label/participants",
            params=_build_query({"limit": limit, "cursor": cursor}),
            omit_firm_id=True,
        )

    def register_participant(
        self,
        request: WhiteLabelParticipantRegistrationRequest,
        *,
        idempotency_key: str,
    ) -> WhiteLabelParticipantOperation:
        key = _white_label_idempotency_key(idempotency_key)
        return self._request(
            "POST",
            "/white-label/participants/registrations",
            json=request,
            extra_headers=_idempotency_headers(key),
            omit_firm_id=True,
            retry_on_failure=True,
        )

    def migrate_participant(
        self,
        request: WhiteLabelParticipantMigrationRequest,
        *,
        idempotency_key: str,
    ) -> WhiteLabelParticipantOperation:
        key = _white_label_idempotency_key(idempotency_key)
        return self._request(
            "POST",
            "/white-label/participants/migrations",
            json=request,
            extra_headers=_idempotency_headers(key),
            omit_firm_id=True,
            retry_on_failure=True,
        )

    def get_participant(self, participant_id: str) -> WhiteLabelParticipant:
        return self._request(
            "GET",
            f"/white-label/participants/{quote(participant_id, safe='')}",
            omit_firm_id=True,
        )

    def request_migration_code(
        self,
        participant_id: str,
        *,
        idempotency_key: str,
    ) -> WhiteLabelMigrationCodeResponse:
        key = _white_label_idempotency_key(idempotency_key)
        return self._request(
            "POST",
            f"/white-label/participants/{quote(participant_id, safe='')}/migration-code",
            extra_headers=_idempotency_headers(key),
            omit_firm_id=True,
            retry_on_failure=True,
        )

    def get_operation(self, operation_id: str) -> WhiteLabelParticipantOperation:
        return self._request(
            "GET",
            f"/white-label/operations/{quote(operation_id, safe='')}",
            omit_firm_id=True,
        )

File: ruby/lib/epostak/resources/white_label.rb
# frozen_string_literal: true

require "erb"

module EPostak
  module Resources
    # Integrator-scoped White Label participant registration and migration.
    class WhiteLabel
      def initialize(http)
        @http = http
      end

      def list_participants(limit: nil, cursor: nil)
        @http.request(
          :get,
          "/white-label/participants",
          query: { limit: limit, cursor: cursor }.compact,
          omit_firm_id: true
        )
      end

      # verificationToken is a one-time FS SR secret. Never log the request.
      def register_participant(request, idempotency_key:)
        post_idempotent("/white-label/participants/registrations", request, idempotency_key)
      end

      # migrationCode is an SMP secret. Never log the request.
      def migrate_participant(request, idempotency_key:)
        post_idempotent("/white-label/participants/migrations", request, idempotency_key)
      end

      def get_participant(participant_id)
        @http.request(
          :get,
          "/white-label/participants/#{ERB::Util.url_encode(participant_id)}",
          omit_firm_id: true
        )
      end

      def request_migration_code(participant_id, idempotency_key:)
        key = idempotency_key!(idempotency_key)
        @http.request(
          :post,
          "/white-label/participants/#{ERB::Util.url_encode(participant_id)}/migration-code",
          idempotency_key: key,
          retry_on_failure: true,
          retry_network_errors: true,
          omit_firm_id: true
        )
      end

      def get_operation(operation_id)
        @http.request(
          :get,
          "/white-label/operations/#{ERB::Util.url_encode(operation_id)}",
          omit_firm_id: true
        )
      end

      private

      def post_idempotent(path, request, idempotency_key)
        key = idempotency_key!(idempotency_key)
        @http.request(
          :post,
          path,
          body: request,
          idempotency_key: key,
          retry_on_failure: true,
          retry_network_errors: true,
          omit_firm_id: true
        )
      end

      def idempotency_key!(value)
        unless value.is_a?(String) && !value.strip.empty? && value.bytesize <= 255
          raise ArgumentError, "White Label idempotency key must be 1-255 UTF-8 bytes"
        end

        value
      end
    end
  end
end
